lib/graphql/analysis/ast/query_complexity.rb

```diff
@@ -8,113 +8,211 @@ class QueryComplexity < Analyzer
-          else
-            raise("Invalid complexity: #{defined_complexity.inspect} on #{field_defn.name}")
+          val = 0
+          @complexities_on_type_by_query.each do |query, complexities_on_type|
+            root_complexity = complexities_on_type.last
+            # Use this entry point to calculate the total complexity
+            total_complexity_for_query = ComplexityMergeFunctions.merged_max_complexity_for_scopes(query, [root_complexity.scoped_children])
+            val += total_complexity_for_query
           end
+          val
         end
 
-        # Selections on an object may apply differently depending on what is _actually_ returned by the resolve function.
-        # Find the maximum possible complexity among those combinations.
-        class AbstractTypeComplexity
-          def initialize
-            @types = Hash.new { |h, k| h[k] = {} }
-          end
-
-          # Return the max possible complexity for types in this selection
-          def max_possible_complexity
-            max = 0
-            @types.each_value do |fields|
-              complexity = fields.each_value.inject(:+)
-              max = complexity if complexity > max
+        # These functions use `ScopedTypeComplexity` objects,
+        # especially their `scoped_children`, to traverse down the tree
+        # and find the max complexity for any possible runtime type.
+        # Yowza.
+        module ComplexityMergeFunctions
+          module_function
+          # When looking at two selection scopes, figure out whether the selections on
+          # `right_scope` should be applied when analyzing `left_scope`.
+          # This is like the `isSubtypeOf` check from the GraphQL specification.
+          # TODO dedup with runtime checks? This is implemented somewhere.
+          def applies_to?(query, left_scope, right_scope)
+            if left_scope == right_scope
+              # This can happen when several branches are being analyzed together
+              true
+            else
+              # Check if these two scopes have _any_ types in common.
+              possible_right_types = query.possible_types(right_scope)
+              possible_left_types = query.possible_types(left_scope)
+              possible_right_types.each do |rt|
+                if possible_left_types.include?(rt)
+                  return true
+                end
+              end
+              false
             end
-            max
           end
 
-          # Store the complexity for the branch on `type_defn`.
-          # Later we will see if this is the max complexity among branches.
-          def merge(type_defn, key, complexity)
-            @types[type_defn][key] = complexity
-          end
-        end
+          def merged_max_complexity_for_scopes(query, scoped_children_hashes)
+            # Figure out what scopes are possible here.
+            all_scopes = []
+            scoped_children_hashes.each do |h|
```

Comment:
`all_scopes = scoped_children_hashes.map(&:keys)`?

Reply:
I also made the implementation decision here based on ✨ imagined ✨  performance. I put it to the test, and it looks to me like a Hash based approach would be best. Thanks for the suggestion, I'll update it!

```ruby
require "benchmark/ips"
require "set"

big_hashes = [
  Hash[("a".."f").each_with_index.map { |k, v| [k, v] }],
  Hash[("e".."j").each_with_index.map { |k, v| [k, v] }],
  Hash[("g".."y").each_with_index.map { |k, v| [k, v] }],
  Hash[("a".."z").each_with_index.map { |k, v| [k, v] }],
]

small_hashes = [
  Hash[("a".."c").each_with_index.map { |k, v| [k, v] }],
  Hash[("b".."f").each_with_index.map { |k, v| [k, v] }],
]

LETTERS = ("a".."f").to_a
def test_include(all_keys)
  LETTERS.each { |l| all_keys.include?(l) }
end

Benchmark.ips do |x|
  x.report("BIG concat") {
    all_keys = []
    big_hashes.each { |h| all_keys.concat(h.keys) }
    all_keys.uniq!
    test_include(all_keys)
  }

  x.report("BIG map/flatten") {
    all_keys = big_hashes.map(&:keys)
    all_keys.flatten!
    all_keys.uniq!
    test_include(all_keys)
  }

  x.report("BIG set") {
    all_keys = Set.new
    big_hashes.each { |h| all_keys.merge(h.keys) }
    test_include(all_keys)
  }

  x.report("BIG hash") {
    all_keys = {}
    big_hashes.each { |h| all_keys.merge!(h) }
    test_include(all_keys)
  }

  x.report("SMALL concat") {
    all_keys = []
    small_hashes.each { |h| all_keys.concat(h.keys) }
    all_keys.uniq!
    test_include(all_keys)
  }

  x.report("SMALL map/flatten") {
    all_keys = small_hashes.map(&:keys)
    all_keys.flatten!
    all_keys.uniq!
    test_include(all_keys)
  }

  x.report("SMALL set") {
    all_keys = Set.new
    small_hashes.each { |h| all_keys.merge(h.keys) }
    test_include(all_keys)
  }

  x.report("SMALL hash") {
    all_keys = {}
    small_hashes.each { |h| all_keys.merge!(h) }
    test_include(all_keys)
  }
end
```

```
~/code/graphql-ruby $ ruby inclusion_benchmark.rb
Warming up --------------------------------------
          BIG concat    17.504k i/100ms
     BIG map/flatten     8.826k i/100ms
             BIG set     7.260k i/100ms
            BIG hash    20.203k i/100ms
        SMALL concat    38.521k i/100ms
   SMALL map/flatten    21.042k i/100ms
           SMALL set    23.320k i/100ms
          SMALL hash    65.533k i/100ms
Calculating -------------------------------------
          BIG concat    184.866k (±10.2%) i/s -    927.712k in   5.088599s
     BIG map/flatten     94.995k (±21.7%) i/s -    450.126k in   5.041971s
             BIG set     71.288k (±11.3%) i/s -    355.740k in   5.078580s
            BIG hash    206.281k (± 9.9%) i/s -      1.030M in   5.059234s
        SMALL concat    504.556k (±13.2%) i/s -      2.465M in   5.027007s
   SMALL map/flatten    291.651k (±11.4%) i/s -      1.431M in   5.001471s
           SMALL set    214.573k (±19.7%) i/s -      1.026M in   5.029503s
          SMALL hash    764.886k (± 7.9%) i/s -      3.801M in   5.017882s
```

Reply:
Awesome! Thanks for the thorough explanation :D